### PR TITLE
tcp: dump shutdown state for unconnected sockets

### PR DESCRIPTION
**Pull request by avagin**
There is no direct way to get this state but we can get it indirectly from polling events for a socket.

### CODE REVIEW

**Comment by adrianreber** [comment]
You probably did not want to commit the strace log file.

**Comment by avagin** [comment]
> You probably did not want to commit the strace log file.

Fixed. Thanks!

File: criu/sk-inet.c [hunk 1]
```diff
@@ -225,6 +226,31 @@ static int dump_sockaddr(union libsoccr_addr *sa, u32 *pb_port, u32 *pb_addr)
 	return -1;
 }
 
+/*
+ * There is no direct way to get shutdown state for unconnected sockets,
+ * but we can get it indirectly from polling events for a socket.
+ */
+static int dump_uncon_sk_shutdown(int lfd, struct inet_sk_desc *sk)
+{
+	struct pollfd pfd = { .fd = lfd, .events = POLLRDHUP | POLLHUP};;
```
Comment (rst0git) [inline criu/sk-inet.c:235]: There is unnecessary `;` at the end of this line.

File: criu/sk-inet.c [hunk 2]
```diff
@@ -225,6 +226,31 @@ static int dump_sockaddr(union libsoccr_addr *sa, u32 *pb_port, u32 *pb_addr)
...
+
+	if (poll(&pfd, 1, 0) != 1) {
+		pr_perror("Unable to poll the socket");
+		return -1;
+	}
+
+	sk->shutdown = 0;
+
+	if (pfd.revents & POLLHUP == 0)
```
Comment (rst0git) [inline criu/sk-inet.c:244]:
```
criu/sk-inet.c:244:18: error: suggest parentheses around comparison in operand of ‘&’ [-Werror=parentheses]
  244 |  if (pfd.revents & POLLHUP == 0)
```

File: criu/sk-inet.c [hunk 3]
```diff
@@ -225,6 +226,31 @@ static int dump_sockaddr(union libsoccr_addr *sa, u32 *pb_port, u32 *pb_addr)
 	return -1;
 }
 
+/*
+ * There is no direct way to get shutdown state for unconnected sockets,
+ * but we can get it indirectly from polling events for a socket.
+ */
```
Comment (rst0git) [inline criu/sk-inet.c:232]: It looks like `ss` can detect `SHUT_RD`, `SHUT_WR`, `SHUT_RDWR` and non-shutdown sockets: https://github.com/shemminger/iproute2/commit/5b816047537bb04620718e798ec09ac9f7a0ad67

File: criu/sk-inet.c [hunk 4]
```diff
@@ -225,6 +226,31 @@ static int dump_sockaddr(union libsoccr_addr *sa, u32 *pb_port, u32 *pb_addr)
...
+static int dump_uncon_sk_shutdown(int lfd, struct inet_sk_desc *sk)
+{
+	struct pollfd pfd = { .fd = lfd, .events = POLLRDHUP | POLLHUP};
+
+	if (poll(&pfd, 1, 0) != 1) {
```
Comment (rst0git) [inline criu/sk-inet.c:237]:
```diff                                                                                                       
-       if (poll(&pfd, 1, 0) != 1) {
+       if (poll(&pfd, 1, 0) != 1 && errno != ENOTCONN) {
                pr_perror("Unable to poll the socket");
                return -1;
        }
```

**Review by rst0git — APPROVED** [review]

**Pull request merged**